File: _BatteryTest_BlueSky/genSOC_v_VTerm.py
import pandas as pd
import time
import matplotlib.pyplot as plt

def generate_SOC_vs_VTerm(file_path, nominal_capacity, mode):
    # Data format inside file (.CSV)
    # {'TimeStamp','TimeDelta','InputCurrent','OutputVoltage'}
    # nominal_capacity - nominal capacity in mAh
    # mode - (1) or (-1) for discharging or charging
    data = pd.read_csv(file_path)
    dt_idx = 1
    in_curr_idx = 2
    out_volt_idx = 3
    n = len(data)

    Cte = 1000/3600 # coulomb_factor 

    # Assume we start from a fully charged battery.
    out_SOC = [1]
    out_VTerm = [data.iloc[1, out_volt_idx]]

    # Do Coulomb counting.
    for i in range(2, n):
        temp_SOC = out_SOC[i - 2] - (Cte*mode*data.iloc[i, dt_idx]*data.iloc[i, in_curr_idx])/nominal_capacity

        if temp_SOC < 0:
            break
        else:
            out_SOC.append(temp_SOC)
            out_VTerm.append(data.iloc[i, out_volt_idx])

    # Create output file.
    plt.plot(out_VTerm, out_SOC)
    plt.title('Terminal Voltage vs SOC')
    plt.legend(['SOC'], loc='upper left')
    plt.xlim([out_VTerm[-2], out_VTerm[0]])
    plt.ylim([0, 1])
    plt.xlabel('Voltage (V)')
    plt.ylabel('SOC')
    plt.show()

    current_time = time.localtime()
    file_name = 'VTerm_v_SOC_' + str(current_time.tm_hour) + str(current_time.tm_min) + '.csv'
    out_table = pd.DataFrame({'VTerm': out_VTerm, 'SOC': out_SOC})
    out_table.to_csv(file_name)

    return True

File: StateOfCharge/genSOC_v_VTerm.m
function result = genSOC_v_VTerm(file, Q, mode)

% Data format inside file (.CSV)
% {'TimeStamp','TimeDelta','InputCurrent','OutputVoltage'}
% Q - nominal capacity in mAh
% mode - (1) or (-1) for dischargin or charging
data = readtable(file);
dt_idx = 2;
inCurr_idx = 3;
outVolt_idx = 4;
n = height(data);

Cte = 1000/3600;

% Assume we start from a fully charged battery.
outSOC = [1];
outVTerm = [data{2, outVolt_idx}];

% Do Coulomb counting.
for i = 3:n
    
    tempSOC = outSOC(i - 2) - (Cte*mode*data{i, dt_idx}*data{i, inCurr_idx})/Q;
    
    if tempSOC < 0 
        break
    else
        outSOC = [outSOC; tempSOC];
        outVTerm = [outVTerm; data{i, outVolt_idx}];
    end
end

% Create output file.

figure
plot(outVTerm, outSOC);
title('Terminal Voltage vs SOC');
legend('SOC', 'Location', 'northwest');
xlim([outVTerm(i-1), outVTerm(1)]);
ylim([0,1]);
xlabel('Voltage (V)');
ylabel('SOC');
hold on

time = clock;
outFile = 'VTerm_v_SOC_' + string(time(4)) + string(time(5)) + '.csv';

outTable = table(outVTerm, outSOC);
writetable(outTable, outFile);



result = true;

end